src/tools/non-promotion-reasons.md

```diff
@@ -1,45 +1,615 @@
+
+{:.good}
+<?code-excerpt "../null_safety_examples/non_promotion/lib/non_promotion.dart (write_combine_ifs)" replace="/else/[!$&!]/g"?>
+{% prettify dart tag=pre+code %}
+void f(bool b, int? i, int? j) {
+  if (i == null) return;
+  if (b) {
+    i = j;
+  } [!else!] {
+    print(i.isEven);
+  }
+}
+{% endprettify %}
+
+In straight-line control flow cases like these (no loops),
+flow analysis takes into account the right hand side of the assignment
+when deciding whether to demote.
+As a result, another way to fix this code is
+to change the type of `j` to `int`.
+
+{:.good}
+<?code-excerpt "../null_safety_examples/non_promotion/lib/non_promotion.dart (write_change_type)" replace="/int j/[!$&!]/g"?>
+{% prettify dart tag=pre+code %}
+void f(bool b, int? i, [!int j!]) {
+  if (i == null) return;
+  if (b) {
+    i = j;
+  }
+  if (!b) {
+    print(i.isEven);
+  }
+}
+{% endprettify %}
+
+
+### Possibly written in a previous loop iteration {#loop-or-switch}
+
+**The cause:**
+You're trying to promote something that
+might have been written to in a previous iteration of a loop,
+and so the promotion was invalidated.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(Link? p) {
+  if (p != null) return;
+  while (true) {    // (1)
+    print(p.value); // (2) ERROR
+    var next = p.next;
+    if (next == null) break;
+    p = next;       // (3)
+  }
+}
+{% endprettify %}
+
+When flow analysis reaches (1),
+it looks ahead and sees the write to `p` at (3).
+But because it's looking ahead,
+it hasn't yet figured out the type of the right-hand side of the assignment,
+so it doesn't know whether it's safe to retain the promotion.
+To be safe, it invalidates the promotion.
+
+You might fix this problem by moving the null check to the top of the loop:
+
+{:.good}
+<?code-excerpt "../null_safety_examples/non_promotion/lib/non_promotion.dart (loop)" replace="/p != null/[!$&!]/g"?>
+{% prettify dart tag=pre+code %}
+void f(Link? p) {
+  while ([!p != null!]) {
+    print(p.value);
+    p = p.next;
+  }
+}
+{% endprettify %}
+
+This situation can also arise in `switch` statements if
+a `case` block has a label,
+because you can use labeled `switch` statements to construct loops:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int i, int? j, int? k) {
+  if (j == null) return;
+  switch (i) {
+    label:
+    case 0:
+      print(j.isEven); // ERROR
+      j = k;
+      continue label;
+  }
+}
+{% endprettify %}
+
+Again, you can fix the problem by moving the null check to the top of the loop:
+
+{:.good}
+<?code-excerpt "../null_safety_examples/non_promotion/lib/non_promotion.dart (switch-loop)" replace="/if .*/[!$&!]/g"?>
+{% prettify dart tag=pre+code %}
+void f(int i, int? j, int? k) {
+  switch (i) {
+    label:
+    case 0:
+      [!if (j == null) return;!]
+      print(j.isEven);
+      j = k;
+      continue label;
+  }
+}
+{% endprettify %}
+
+
+
+### In catch after possible write in try
```

Comment:
Perhaps another simplified anchor may be useful in this case as well.

Reply:
Good catch, I'll do that before merging.